### ISSUE TITLE
openGL blocking under kubuntu: possible solution /  workaround

### ISSUE THREAD

**Issue by jnspitale** [hannorein/rebound#516]
This is a followup to closed issue #284: "GLSL 3.30 not supported, examples core dump."  I'm running kubuntu under unbuntu 18.04 with an intel graphics card and an nvidia card.  The issue is the same with either card.  

As in the referenced thread, with OPENGL=1, I just get a window frame displaying the underlying contents of the desktop using the shearing sheet example, while the solar system example works fine.  The problem apparently has to do with the synchronization of the display and simulation threads, with the display thread getting locked out.  When I comment out all of the thread management in the simulation loop in reb_integrate_raw, the example works fine, and the solar system example continues to work.  To be precise:

```
static void* reb_integrate_raw(void* args){
    reb_sigint = 0;
    signal(SIGINT, reb_sigint_handler);
    struct reb_thread_info* thread_info = (struct reb_thread_info*)args;
	struct reb_simulation* const r = thread_info->r;
#ifdef MPI
    // Distribute particles
    reb_communication_mpi_distribute_particles(r);
#endif // MPI

    double last_full_dt = r->dt; // need to store r->dt in case timestep gets artificially shrunk to meet exact_finish_time=1
    r->dt_last_done = 0.; // Reset in case first timestep attempt will fail

    r->status = REB_RUNNING;
    reb_run_heartbeat(r);
    while(reb_check_exit(r,thread_info->tmax,&last_full_dt)<0){
#ifdef OPENGL
//        if (r->display_data){
//            if (r->display_data->opengl_enabled){ pthread_mutex_lock(&(r->display_data->mutex)); }
//        }
#endif // OPENGL
        if (r->simulationarchive_filename){ reb_simulationarchive_heartbeat(r);}
        reb_step(r); 
        reb_run_heartbeat(r);
        if (reb_sigint== 1){
            r->status = REB_EXIT_SIGINT;
        }
#ifdef OPENGL
//        if (r->display_data){
//            if (r->display_data->opengl_enabled){ pthread_mutex_unlock(&(r->display_data->mutex)); }
//        }
#endif // OPENGL
    }

    reb_integrator_synchronize(r);
    if (r->display_heartbeat){                          // Display Heartbeat
        r->display_heartbeat(r); 
    }
    if(r->exact_finish_time==1){ // if finish_time = 1, r->dt could have been shrunk, so set to the last full timestep
        r->dt = last_full_dt; 
    }
    if (r->simulationarchive_filename){ reb_simulationarchive_heartbeat(r);}

    return NULL;
}
```

Hopefully this helps find a more general solution to the problem.

-joe

**Comment by hannorein** [comment]
Thanks for sharing this. However, I don't think this is the right thing to do. Accessing the same memory from multiple threads will at some point lead to unexpected behaviour. Also, if the issue is really the synchronization, then it shouldn't work for the solar system example either. I suspect some part of the solid sphere rendering pipeline is not supported on your operating system, graphic card, or glfw library version. You could try changing the display style to points before starting the simulation (`reb_display_data->spheres=0`). Ideally, there should be a more graceful exit with a meaningful message. However, I'm not checking for any errors in the OpenGL API.

In the long run, I don't think the OpenGL visualization is maintainable, especially now that OpenGL is deprecated on MacOS. I'm not sure what to replace it with. There seems to be no good cross-platform solution that ticks all the boxes.

**Comment by jnspitale** [comment]
Yeah, that's why I think it's more of a work-around than a solution.  Changing the display style to points just seg faults with and without the above modifications.  I agree with your concern of unexpected results, but this is still useful for development.  I wouldn't want to run it like this for a production simulation, but the visualizations aren't needed in that case.

-joe

**Comment by hannorein** [comment]
I agree. Let's leave the issue open. If I find some time, I'll try to look into it. It might just not happen right away!

**Comment by hannorein** [comment]
I've tried to look into this. Everything works on my Mac. And I don't have a linux computer at home. My attempt to get OpenGL working with X forwarding over ssh failed. And getting hardware acceleration working on a virtual machine might also not be very straightforward. So I'm afraid this has to wait until I have physical access to a linux computer again.

**Comment by jnspitale** [comment]
Thanks for looking into it.  It doesn't seem like a high priority issue.

-joe

**Comment by hannorein** [comment]
An improved visualization module has been implemented in version 4.0.0.
